experiment/httphostheader/httphostheader.go

```diff
@@ -0,0 +1,78 @@
+// Package httphostheader contains the HTTP host header network experiment.
+//
+// This experiment has not been specified yet. It is nonetheless available for testing
+// and as a building block that other experiments could reuse.
+package httphostheader
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"github.com/ooni/probe-engine/experiment/urlgetter"
+	"github.com/ooni/probe-engine/model"
+)
+
+const (
+	testName    = "http_host_header"
+	testVersion = "0.1.0"
+)
+
+// Config contains the experiment config.
+type Config struct {
+	// TestHelperURL is the address of the test helper.
+	TestHelperURL string
+}
+
+// TestKeys contains httphost test keys.
+type TestKeys struct {
+	urlgetter.TestKeys
+	THAddress string `json:"th_address"`
+}
+
+// Measurer performs the measurement.
+type Measurer struct {
+	config Config
+}
+
+// ExperimentName implements ExperimentMeasurer.ExperiExperimentName.
+func (m *Measurer) ExperimentName() string {
+	return testName
+}
+
+// ExperimentVersion implements ExperimentMeasurer.ExperimentVersion.
+func (m *Measurer) ExperimentVersion() string {
+	return testVersion
+}
+
+// Run implements ExperimentMeasurer.Run.
+func (m *Measurer) Run(
+	ctx context.Context,
+	sess model.ExperimentSession,
+	measurement *model.Measurement,
+	callbacks model.ExperimentCallbacks,
+) error {
+	if measurement.Input == "" {
+		return errors.New("experiment requires input")
+	}
+	if m.config.TestHelperURL == "" {
+		m.config.TestHelperURL = "http://www.example.com"
+	}
+	urlgetter.RegisterExtensions(measurement)
+	g := urlgetter.Getter{
+		Begin: measurement.MeasurementStartTimeSaved,
+		Config: urlgetter.Config{
+			HTTPHost: string(measurement.Input),
+		},
+		Session: sess,
+		Target:  fmt.Sprintf(m.config.TestHelperURL),
+	}
+	tk, _ := g.Get(ctx)
+	measurement.TestKeys = tk
+	return nil
```

Comment:
Here I've changed the code to return `nil` rather than `g.Get`'s error. Cases of censorship where the connection is reset cause `g.Get` to return an error. But the caller may not submit the measurement if we return `err` here. Thus, always returning `nil` is a much better choice here.

This semantics was not documented. So, as part of the same commit, I have also ensured that it's documented otherwise contributors, who cannot read my mind, are not aware of this fact 🙈.

(This semantics may not be the best one, and it may make sense to always submit the result of a measurement. I should probably open an issue about this to spend some more time thinking about this topic and then, possibly, bulk changing all the experiments to propagate errors back.)

Reply:
Issue here: https://github.com/ooni/probe-engine/issues/1017